### PR TITLE
Add multi_array option to timeseries_dataset_from_array

### LINKED ISSUE #16055: Generalize timeseries_dataset_from_array to allow sampling from multiple arrays
**System information**.

TensorFlow version (you are using): 2.8.0
Are you willing to contribute it (Yes/No) : Yes

**Describe the feature and the current behavior/state**.

I have a dataset consisting of time series for multiple stocks and need to draw sample windows from all of them simultaneously. The current implementation only allows drawing from a single time series.

**Will this change the current api? How?**

Yes, additional parameter `multi_array` with default value False.

**Who will benefit from this feature?**

Anyone with a need to draw mixed sample windows from a collection of time series.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no): Yes
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

Solution calculates `start_positions` for each array separately. It then concatenates the individual arrays (padding or truncating the `targets`) and combines the `start_positions` with offsets to index into the new single `data` and `targets` arrays. From there it proceeds unchanged.

### PR DESCRIPTION
**Pull request by hotchipsveg**
This PR implements the change proposed in https://github.com/keras-team/keras/issues/16055

It allows creating a dataset drawing sample batches from multiple time series simultaneously.

This is useful when dealing with data consisting of multiple related time series, e.g. price series for a basket of stocks, or a time series with gaps, where the sample windows should not cross the gap boundaries.

Closes #16055

### CODE REVIEW

**Comment by fchollet** [comment]
Thanks for the PR.

> This is useful when dealing with data consisting of multiple related time series, e.g. price series for a basket of stocks

Isn't this better achieved by simply concatenating the different timeseries into a multidimensional timeseries and using that as the input to this utility?

**Comment by hotchipsveg** [comment]
Thanks for the comment.

The individual series will in general have different lengths in the time dimension, so won't fit into a square array. Also, the model should not be tied to the particular shape this multidimensional array would have. In the stock example, stocks will be added and dropped over time. Also, the model should be generic over the stock dimension, i.e not make different predictions depending on the stock index.

Wouldn't concatenating the series before passing them to the `timeseries_dataset_from_array` function, as you suggest, require overriding model methods to adapt it to the dataset in order to achieve these objectives? The goal of the PR is to be able to call the standard fit and predict methods.

**Comment by fchollet** [comment]
> The individual series will in general have different lengths in the time dimension, so won't fit into a square array

Not sure I follow -- how do you intend for a model to process the data if it doesn't fit into a contiguous array?

**Comment by hotchipsveg** [comment]
For example, if I have 10 days worth of data and want to sample windows of length 3. Series A has data for all days, series B only for the first 5, and C is missing day 4, so I have series of lengths 10, 5, 3 and 6 to draw window samples from. Combining them into a multdimensional array and passing this to the existing function will not work, but

**Comment by hotchipsveg** [comment]
> Not sure I follow -- how do you intend for a model to process the data if it doesn't fit into a contiguous array?

That's exactly what the PR would accomplish. It would allow me to draw samples from data that doesn't fit into a contiguous array and feed them into a model that isn't aware of the fragmentation of the data.

**Comment by fchollet** [comment]
> I don't see how combining the series into a multdimensional array and passing this to the existing function could work.

You'd use padding.

> It allows me to draw samples from data that doesn't fit into a contiguous array and feed them into a model that isn't aware of the fragmentation of the data.

If you're not using padding, then either your timestep will be irregular, or your input dimension will be irregular. Which is it?

**Comment by hotchipsveg** [comment]
Both input dimension (number of series) and time dimension are irregular. Passing padded multidimensional arrays to the function will give me samples of shape (batch, time, series, features) so I would have to :
i) flatten out the series dimension
ii) discard data with pad values (don't want the default masking behaviour of skipping/ignoring).
The multi_array option gives me clean window samples of shape (batch, time, features) that I can feed directly into an LSTM layer.

**Comment by fchollet** [comment]
Ok, thanks for the clarification.

Overall, this is a niche feature, and we do not believe we should add it to the core API. Since it is useful to you, you can simply keep it as part of your own add-on library. If it is made publicly available, others with the same use case might also benefit from it.

From our [API design guidelines](https://github.com/keras-team/governance/blob/master/keras_api_design_guidelines.md#carefully-weigh-whether-a-new-feature-should-be-included):

> [any new feature] should be broadly useful to our users, rather than a niche feature that is only relevant to a specific vertical of researchers. Niche features should be maintained independently by those who need them (e.g. by extending the API via subclassing), as third-party add-on packages.
